datadog_checks_base/datadog_checks/base/checks/base.py

```diff
@@ -945,6 +966,31 @@ def service_check(self, name, status, tags=None, hostname=None, message=None, ra
             self, self.check_id, self._format_namespace(name, raw), status, tags, hostname, message
         )
 
+    def send_log(self, data, checkpoint=''):
+        # type: (dict[str, str], str) -> None
+        """Send a log for submission.
+
+        Parameters:
+
+            data (dict[str, str]):
+                the log data to send
+            checkpoint (str):
+                metadata associated with the log which will be saved to disk. the most recent value may be
+                retrieved with the `get_log_checkpoint` method.
+        """
+        attributes = data.copy()
+        if self.formatted_tags and 'ddtags' not in attributes:
+            attributes['ddtags'] = self.formatted_tags
+
+        datadog_agent.send_log(to_json(attributes), self.check_id)
+        if checkpoint:
+            self.write_persistent_cache('log_checkpoint', checkpoint)
+
+    def get_log_checkpoint(self):
+        # type: () -> str
+        """Returns the most recent log checkpoint from disk."""
+        return self.read_persistent_cache('log_checkpoint')
+
```

Comment:
So in a check I'd write something like this?
```python
def check(self):
    data = self.create_data() # the business logic, assume it's defined
    self.send_log(data, checkpoint=self.get_log_checkpoint())
```
Is there ever a case where we wouldn't want to provide the checkpoint?

Reply:
On the first run (possibly in the check initialization functions) you would use the latest checkpoint yes but every time after that you would supply a new checkpoint. What the checkpoint is would be completely implementation-specific.

Reply:
1. What happens when multiple instances of the same integration are submitting logs with checkpoints? All instances will submit the same checkpoint?
2. Is there ever a case when we **don't want to use checkpoints**? I'm worried this interface doesn't encourage ppl to "do the right thing". I.e. partner devs **have to remember** to add a checkpoint and we as reviewers **have to remember** to bring this up during review. I don't trust our collective memories. Below is a sketch for an API that submits the checkpoint by default.



```python
class AgentCheck:
    def __init__(self, *args, **kwargs):
        ...
        self._logs_checkpoint = self.compute_logs_checkpoint()
        ...
    def compute_logs_checkpoint(self) -> Optional[str]:
        '''
        Some reasonable default like using the class name or combining several constants.
        User can override this with their custom logic. Return None to disable.
        '''
    def send_log(self, data):
         '''All the same with `self._logs_checkpoint` instead of `checkpoint`'''
```

Reply:
1. Can you please explain what you mean? I don't quite understand the concern here. Each instance would be collecting from a different source and therefore would save different checkpoints.
2. What is saved as the checkpoint in your example?

Reply:
I guess both questions boil down to "what is a checkpoint?".

Assuming we roll with the existing implementation, here's an example concrete class that **plays by the rules**:

```
class LogsCheck(AgentCheck):
    def __init__(self, *args, **kwargs):
        ... # do other stuff, probably invoke super() etc
        self.checkpoint = self.get_log_checkpoint()
        if self.checkpoint is None and logs_enabled:
            self.checkpoint = "abracadabra"
    def check(self, *args):
        ... # collect data
        self.send_log(data, self.checkpoint)
```

Now given this definition how do we differentiate between multiple instances of the check?

Reply:
I still don't quite understand the question so let me answer your question first: a checkpoint is used to not submit duplicate logs. For example, say one is submitting logs from data provided by a paginated JSON REST endpoint and you are on item 5 from page 3, you would want to save the page number and potentially the item number as well.

Reply:
Thanks, that helps.
Back to my question: how do we make checkpointing **opt-out**, not **opt-in**? 

One idea would be to make `checkpoint` a required argument with a special value `None` that disables it. 

wdyt?

Reply:
This I kept as optional because not everything is a crawler but as a compromise the cursor is now required for the new crawler structure `datadog_checks.base.checks.logs.crawler.stream.LogRecord`